src/dataflow/src/source/mod.rs

```diff
@@ -657,31 +657,25 @@ impl SourceTransaction<'_> {
 /// transaction.
 ///
 /// When a transaction is started using [start_tx](Timestamper::start_tx) the internal clock will be
+/// // This send is guarded by the `timestamp` LockGuard, so we know its accurate:
+/// the
 /// frozen and any subsequent rows will be timestamped with the exact same timestamp. The
 /// transaction is committed automatically as soon as the transaction object gets dropped.
-pub struct Timestamper {
-    inner: Arc<RwLock<Timestamp>>,
+pub struct TxnTimestamper {
+    timestamp: Arc<RwLock<Timestamp>>,
     sender: EventSender,
-    tick_duration: Duration,
-    now: NowFn,
 }
 
-impl Timestamper {
-    fn new(sender: EventSender, tick_duration: Duration, now: NowFn) -> Self {
-        let ts = now();
-        Self {
-            inner: Arc::new(RwLock::new(ts)),
-            sender,
-            tick_duration,
-            now,
-        }
-    }
-
+impl TxnTimestamper {
     /// Start a transaction at a particular point in time. The timestamper will freeze its internal
     /// clock while a transaction is active.
+    ///
+    /// NOTE: Calling `start_tx` a second time, or other methods on `Timestamper`,
+    /// before dropping the transaction guard could result in a deadlock.
```

Comment:
This could be hard, depending on where this thing gets stashed, but if it took `&mut self` it'd be impossible to trigger that class of error!

Reply:
good point, that seemed to dropped in RELATIVELY easily, mode the rest of the code cleaner too